### ISSUE TITLE
Angular 1.4.6 sporadic $http errors with status -1

### ISSUE THREAD

**Issue by jeffwhelpley** [angular/angular.js#12920]
@petebacondarwin I apologize ahead of time for not being more specific, but this is an issue I can't reproduce in dev and I only see in our prod error logging when our app us under load.

Sporadically on Android and Windows (mostly Android) with Chrome (only happens with Chrome), our Angular app that was just upgraded to 1.4.6 gets a -1 response code. There is no error and the request looks fine. In fact, when I retry the transaction after one of these errors, it goes through just fine. When I revert back to Angular 1.4.5 the errors go away.

I really apologize, but I can't tell you how to reproduce this. I was hoping that if I post this someone else will run into the same issue and be able to narrow in on the issue much better than I have.

The code that I am using when this error occurs is right here:

https://github.com/gethuman/pancakes-angular/blob/master/lib/ngapp/ajax.js#L164

The stack trace for this is unique to my code since I am explicitly throwing it when $http returns -1 status code, but it looks like this:

```
Error: error httpCode -1 headers undefined conf {"method":"GET","transformRequest":[null],"transformRespons…
  at completeRequest(angular.js:10668:8)
  at completeRequest(angular.js:10717:6)
  at callback(angular.js:10545:35)
  at $apply(angular.js:16030:12)
  at $digest(angular.js:15733:14)
  at $eval(angular.js:15922:15)
  at ?(angular.js:14678:27)
  at fn(angular.js:10258:12)
  at ?(/js/gh.common.1442944366792.js:1132:1973)
```

**Comment by petebacondarwin** [comment]
This will be related to this commithttps://github.com/angular/angular.js/commit/38520a1a73fffb6cfeffc7edfcab5be33e1619eb

**Comment by pkozlowski-opensource** [comment]
I'm surprised that 38520a1 landed as part of 1.4.6 since it is a (kind of) breaking change. Anyway, -1 indicates a canceled / timed-out request and this would explain sporadic nature of those errors. 

Once again 38520a1 is a breaking change so @jeffwhelpley you might need to adjust your code to take -1 into account (it was previously normalized to 0 so your checks need to be adjusted as well)

**Comment by Narretz** [comment]
We can still revert it in 1.4.x.
Also, shouldn't the commit message include a BC note if it might be a BC?

**Comment by jeffwhelpley** [comment]
Ah, yes, I can see how this would be an issue. Right now I have a re-try mechanism for cancelled / timed-out requests in place using $httpProvider.interceptors, but looks for status = 0. I will adjust my code so it will look for -1 as well.

**Comment by petebacondarwin** [comment]
Glad you have it sorted @jeffwhelpley.

I am not convinced that this is a real breaking change. Previously, failed requests (e.g. timeouts) were being reported by Angular as successful. In @jeffwhelpley's case his app would have been receiving a response that contained no data - or if it was a POST request, then his app would be believing that such a request had succeeded when in fact it may have actually failed.

This fix now correctly reports such failures back to the app, which can then respond appropriately.

**Comment by vitaly-t** [comment]
Where is the fix for this to have this issue closed?

What about 1.3.x? I saw the same change getting into 1.3.19 with the same consequence.

**Comment by jeffwhelpley** [comment]
So, previously an http timeout was returning a status code of 0 and I was doing something like this:

```
if (!response.status) {
  // do retry
}
```

With the changes, you now have to look for -1 since that also means a timeout. So, I have:

```
if (!response.status || response.status === -1) {
  // do retry
}
```

**Comment by petebacondarwin** [comment]
Or simply:

```js
if (response.status <= 0) {
  // do retry
}
```

**Comment by doitlikepruett** [comment]
Hi there, I'm still seeing this issue in 1.6.8. Exactly as described above, sporadic, statuscode -1, unable to reproduce in dev. Often a series of posts fail right in a row, only milliseconds away from each other. Requests that happen right before and after are fine. Sometimes there's ~20 seconds before failure, sometimes in <5 ms. @jeffwhelpley Besides retry is there anything else you can share as to what was causing the issue for you? I've been on this a week so any assistance is much appreciated.

**Comment by jeffwhelpley** [comment]
No, sorry @doitlikepruett. This was quite a while ago and I haven't seen this issue since. Sorry I can't help more.

**Comment by jo-soft** [comment]
@doitlikepruett we have this issue frequently. usually it seems to be related to timeouts but it may also appear due to an cors-issue.

**Comment by doitlikepruett** [comment]
@jo-soft Please share what you find if you work it out. We are getting around most of these right now by implementing an auto retry that will repeat the call up to 5x for these types of failures.

**Comment by jo-soft** [comment]
@doitlikepruett Somehow angularJS is eating errorMessages. If  you do a simple XHR request at least in the  console you can see more detailed error messages.
Unfortunately I did not find a way to log them.
